File: 1992.py
class Solution:
    def findFarmland(self, land: List[List[int]]) -> List[List[int]]:
        def find_bounds(farm):
            return [min(square[0] for square in farm), min(square[1] for square in farm), max(square[0] for square in farm), max(square[1] for square in farm)]
        farms = []

        # Essentially a copy of #200
        m = len(land)
        n = len(land[0])

        def complete_island(i,j):
            checked = set()
            new = set()
            new.add((i,j))
            while bool(new):
                point = new.pop()
                checked.add(point)
                point_x = point[0]
                point_y = point[1]
                for p in range(max(point_x-1, 0), min(point_x+2, m)):
                    for q in range(max(point_y-1, 0), min(point_y+2, n)):
                        if land[p][q] == 1 and (point_x-p)*(point_y-q) == 0 and (point_x != p or point_y != q):
                            if (p,q) not in checked:
                                new.add((p,q))
            return checked

        num_islands = 0
        for x in range(m):
            for y in range(n):
                if land[x][y] == 1:
                    num_islands += 1
                    island = complete_island(x,y)

                    farms.append(find_bounds(island))
                    # Wipe everything in this island
                    for point in island:
                        land[point[0]][point[1]] = 0
        return farms

File: code/1992.cpp
class Solution {
public:
    vector<vector<int>> findFarmland(vector<vector<int>>& land) {
        int m = land.size();
        int n = land[0].size();
        vector<vector<int>> ans;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (land[i][j] == 1) {
                    int x1 = i;
                    int y1 = j;
                    int x2 = i;
                    int y2 = j;
                    while (x2 < (m - 1) && land[x2+1][y1] == 1) {
                        x2++;
                    }
                    while (y2 < (n - 1) && land[x1][y2+1] == 1) {
                        y2++;
                    }
                    vector<int> cand = {x1, y1, x2, y2};
                    ans.push_back(cand);
                    for (int a = x1; a <= x2; a++) {
                        for (int b = y1; b <= y2; b++) {
                            land[a][b] = 0;
                        }
                    }
                }
            }
        }
        return ans;
        
    }
};
